#7803 Campaigns back-end

CarlosNZ opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #7803

# 👩🏻‍💻 What does this PR do?

This was pretty much done by @andreievg. I was going to include it with the front-end PR, but I think it'll be needed for other issues in https://github.com/msupply-foundation/open-msupply/issues/7694, so here it is.

Adds "Campaigns" table, plus required database/GraphQL queries and mutations

## 💌 Any notes for the reviewer?

Only way to test for now is using GraphiQL explorer

# 🧪 Testing

<!-- Explain the steps you'd take to test the changes of this PR manually -->

- [ ] Test with GraphiQL explorere
- [ ] Wait for front-end to test more thoroughly


# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

jmbrunskill commented on server/repository/src/db_diesel/campaign/campaign_row.rs:
```diff
@@ -0,0 +1,107 @@
+        id -> Text,
+        name -> Text,
+        start_date -> Nullable<Date>,
+        end_date -> Nullable<Date>,
```
I think we should add deleted_datetime and implement soft delete.
We'll still want to associate campaigns with old stock but won't want to see them.
Arguably we can just change the end date to hide it but still think deleted_datetime would be good.

jmbrunskill commented on server/repository/src/db_diesel/campaign/campaign_row.rs:
```diff
@@ -0,0 +1,107 @@
+        Ok(result)
+    }
+
+    pub fn delete(&self, campaign_id: &str) -> Result<i64, RepositoryError> {
```
soft delete?

jmbrunskill reviewed (changes requested):
Can you fix postgres tests please (just need to add to get_dates() for postgres I think)
Otherwise looks pretty good, would be keen to implement soft delete but happy to merge as is and come back to that if we need it.

CarlosNZ commented:
Okay @jmbrunskill, I've added the soft-delete and fixed refresh_dates tests if you want to check it?

Also -- there is sync test for campaigns that Andrei had created, but it wasn't imported and run. I tried including it, but it's giving another error that I don't quite understand. You can see the commented out imports in `/Users/carl/GitHub/open-msupply/server/service/src/sync/test/test_data/mod.rs` if you want to have a look, otherwise I'll leave it till after the front-end API if that's cool?

jmbrunskill commented:
> Okay @jmbrunskill, I've added the soft-delete and fixed refresh_dates tests if you want to check it?
> 
> Also -- there is sync test for campaigns that Andrei had created, but it wasn't imported and run. I tried including it, but it's giving another error that I don't quite understand. You can see the commented out imports in `/Users/carl/GitHub/open-msupply/server/service/src/sync/test/test_data/mod.rs` if you want to have a look, otherwise I'll leave it till after the front-end API if that's cool?

Pushed a fix, which hopefully works for you too.
It looks like the item_variant and campaign records were using the same ID, Not 100% sure why that is a problem but should be ok now that they're not duplicated.

jmbrunskill commented:
Also pushed a fix for another test failure due to duplicate test_database names

jmbrunskill reviewed (approved):
Thanks Carl, I think this is good to go now.
